src/coreComponents/finiteElement/kernelInterface/KernelBase.hpp

```diff
@@ -466,6 +467,114 @@ real64 regionBasedKernelApplication( MeshLevel & mesh,
 
   return maxResidualContribution;
 }
+
+/**
+ * @brief Performs a loop over specific regions (by type and name) and calls a kernel launch on the subregions
+ *   with compile time knowledge of sub-loop bounds such as number of nodes and quadrature points per element.
+ * @tparam POLICY The RAJA launch policy to pass to the kernel launch.
+ * @tparam DISPATCH_TYPE_LIST a TypeList where each entry is also TypeList and contains a tuple of the form
+ *   (SUBREGION_TYPE, CONSTITUTIVE_TYPE, FE_TYPE). This is a list of kernel instantiations to dispatch over.
+ * @tparam KERNEL_FACTORY The type of @p kernelFactory, typically an instantiation of @c KernelFactory, and
+ *   must adhere to that interface.
+ * @param mesh The MeshLevel object.
+ * @param targetRegions The names of the target regions(of type @p SUBREGION_TYPE) to apply the @p KERNEL_TEMPLATE.
+ * @param finiteElementName The name of the finite element.
+ * @param constitutiveStringName The key to the constitutive model name found on the Region.
+ * @param kernelFactory The object used to construct the kernel.
+ * @return The maximum contribution to the residual, which may be used to scale the residual.
+ *
+ * @details Loops over all regions Applies/Launches a kernel specified by the @p KERNEL_TEMPLATE through
+ * #::geos::finiteElement::KernelBase::kernelLaunch().
+ */
+template< typename POLICY,
+          typename DISPATCH_TYPE_LIST,
+          typename KERNEL_FACTORY >
+static
+real64 regionBasedKernelApplication( MeshLevel & mesh,
+                                     arrayView1d< string const > const & targetRegions,
+                                     string const & finiteElementName,
+                                     string const & constitutiveStringName,
+                                     KERNEL_FACTORY & kernelFactory )
+{
+  GEOS_MARK_FUNCTION;
+  // save the maximum residual contribution for scaling residuals for convergence criteria.
+  real64 maxResidualContribution = 0;
+
+  NodeManager & nodeManager = mesh.getNodeManager();
+  EdgeManager & edgeManager = mesh.getEdgeManager();
+  FaceManager & faceManager = mesh.getFaceManager();
+  ElementRegionManager & elementRegionManager = mesh.getElemManager();
+
+  // using SUBREGION_TYPES = types::Slice< DISPATCH_TYPE_LIST >;
+  using SUBREGION_TYPES = types::TypeList< CellElementSubRegion >;
```

Comment:
This is a hack, but I am not sure how we want to do this. The `types::Slice` that I commented out will return a list of the first element from the list-of-lists `DISPATCH_TYPE_LIST`, but for what we have now this is just a list containing many copies of `CellElementSubRegion`. So many copies in fact that NVCC gives up in `forElementSubRegions`.

Reply:
Can we just extract a list of unique subregion types from that slice? I am not saying I know how to do it but I am sure we can come up with some meta-programming magic to do that.  To be seen if nvcc does not crash while doing that though...

Reply:
I thought about that, I wanted to check in here before I invested the time.

Reply:
we could also just loop over all `subRegions`  and then have a runtime check on the type inside the dispatch and only launch the kernels if the real type matches the dispatched one. Basically if the subregion cast failsinstead of errorring out we just ignore it.

Reply:
That was what I tried at first. Aside from the fact that I imagine there are cases where we would prefer to error out instead of silently proceeding it has the problem that not all subregions have the `FiniteElementBase` subgroup so the following call will fail

```c++
FiniteElementBase & subRegionFE = elementSubRegion.template getReference< FiniteElementBase >( finiteElementName );
```

If we want to go this way we could do the same thing as we do for the constitutive model and allocate a `NullFiniteElement` class.

Reply:
I had not though of that... this is certainly not the most elegant path so let's see if we can come up with something cleaner as we add use cases for this.

Reply:
Certainly for now the simplest thing would be to just keep it as is `using SUBREGION_TYPES = types::TypeList< CellElementSubRegion >;` and then re-evaluate once we have a need for other types.